src/satosa/micro_services/custom_routing.py

```diff
@@ -2,14 +2,117 @@
+    """
+    SATOSA exception raised by CustomRouting rules
+    """
+    pass
+
+
+class DecideBackendByTargetIdP(RequestMicroService):
+    """
+    Select which backend should be used based on who is the SAML IDP
+    """
+
+    def __init__(self, config:dict, *args, **kwargs):
+        """
+        Constructor.
+        :param config: microservice configuration loaded from yaml file
+        :type config: Dict[str, Dict[str, str]]
+        """
+        super().__init__(*args, **kwargs)
+        self.target_mapping = config['target_mapping']
+        self.endpoint_paths = config['endpoint_paths']
+        self.default_backend = config['default_backend']
+
+        if not isinstance(self.endpoint_paths, list):
+            raise SATOSAConfigurationError()
+
+    def register_endpoints(self):
+        """
+        URL mapping of additional endpoints this micro service needs to register for callbacks.
+
+        Example of a mapping from the url path '/callback' to the callback() method of a micro service:
+            reg_endp = [
+                ("^/callback1$", self.callback),
+            ]
+
+        :rtype List[Tuple[str, Callable[[satosa.context.Context, Any], satosa.response.Response]]]
+
+        :return: A list with functions and args bound to a specific endpoint url,
+                 [(regexp, Callable[[satosa.context.Context], satosa.response.Response]), ...]
+        """
+
+        # this intercepts disco response
+        return [
+            (path , self.backend_by_entityid)
+            for path in self.endpoint_paths
+        ]
+
+    def _get_backend(self, context:Context, entity_id:str) -> str:
+        """
+        returns the Target Backend to use
+        """
+        return (
+            self.target_mapping.get(entity_id) or
+            self.default_backend
+        )
+
+    def process(self, context:Context, data:dict):
+        """
+        Will modify the context.target_backend attribute based on the target entityid.
+        :param context: request context
+        :param data: the internal request
+        """
+        entity_id = context.request.get('entityID')
+        if entity_id:
+            self._rewrite_context(entity_id, context)
+        return super().process(context, data)
+
+    def _rewrite_context(self, entity_id:str, context:Context) -> None:
+        tr_backend = self._get_backend(context, entity_id)
+        context.decorate(Context.KEY_TARGET_ENTITYID, entity_id)
+        context.target_frontend = context.target_frontend or context.state.get('ROUTER')
+        native_backend = context.target_backend
+        msg = (f'Found DecideBackendByTarget ({self.name} microservice) '
+               f'redirecting {entity_id} from {native_backend} '
+               f'backend to {tr_backend}')
+        logger.info(msg)
+        context.target_backend = tr_backend
+
+    def backend_by_entityid(self, context:Context):
+        entity_id = context.request.get('entityID')
```

Comment:
based on the discussion on #366 
- IdpHinting should run before DecideBackendByTargetIdP
- as the name says DecideBackend "ByTargetIdP" we should be using the target-idp as this has been set by IdpHinting or other micro-services

```suggestion
        entity_id = context.get_decoration(Context.KEY_TARGET_ENTITYID)
```

Reply:
done here:
dcda00627e46f56a1af7300a3131280d3405fa85#diff-643d6b17c79aec1dad0acafabfca8efe6dd3dcca590bcfda363aa44b06079830R63